### PR TITLE
Store the TransactionHash and Index On FundingInfo

### PR DESCRIPTION
**Pull request by beaushinkle**
We also correctly calculate the transaction hash by reversing the first 32 bytes of the outpoint before encoding them in hex, since the outpoint stores them in little-endian.

This one is a little trickier to test, since not-reversing the bytes also passes our validators and we hard-code our funding info for unit tests. With that in mind, here's the whole flow:

First, pull down the branch for https://github.com/keep-network/local-setup/pull/71 (or merge that first), since we need the `getBalance` and `sendRawTransaction` commands it provides.

Generate 3 new bitcoin addresses via:
```
$ bitcoind-wallet getNewAddress
```

Edit `keep-ecdsa/configs/config.local.1.toml`, `keep-ecdsa/configs/config.local.2.toml`, and `keep-ecdsa/configs/config.local.3.toml` to have:

```
[Extensions.TBTC]
	TBTCSystem = "<address from [ethereum.ContractAddresses.TBTCSystem]"

  [Extensions.TBTC.BTCRefunds]
  	BeneficiaryAddress = "<address from bitcoind-wallet getNewAddress> "
  	MaxFeePerVByte = 75
	BitcoinChainName = "regtest"

[TSS]
	PreParamsTargetPoolSize = 2 # not strictly required, but makes the whole thing faster
```

Run keep-core and the 3 keep ecdsa nodes

Create a new deposit from `local-setup/tbtc.js`:
```
$ bin/tbtc.js --rpc ws://127.0.0.1:8546 deposit new 10000000
```
Watch for a message like:
```
Got deposit address: <funding address> ; fund with: 10000000 satoshis please.
```
Then, fund that address:
```
$ bitcoind-wallet sendToAddress <funding address> 0.1
```

Wait for it to confirm with a message like:
```
Deposit is active, minting...
[object Promise]	99950000000000000
```

Now, we have to undercollateralize the deposit from the `local-setup/tbtc/solidity/scripts` folder:
```
$ truffle exec multiply-local-eth-btc-price.js 0.7
```
Once that completes, you can check the active deposits from `local-setup/tbtc.js`:
```
$ bin/tbtc.js --rpc ws://127.0.0.1:8546 deposit list --vending-machine

(node:28834) ExperimentalWarning: Importing JSON modules is an experimental feature. This feature could change at any time
(Use `node --trace-warnings ...` to show where the warning was created)
<deposit address>	                        ACTIVE	10000000	104
```

And then liquidate it:

```
$ bin/tbtc.js --rpc ws://127.0.0.1:8546 deposit <deposit address> liquidate
```

Now, find one of your `keep-ecdsa` nodes, and monitor it for some logs that look like this (it may take a while):
```
2021-04-16T11:41:08.425-0400	WARN	keep-ecdsa	Please broadcast Bitcoin transaction <transaction-hex>
2021-04-16T11:41:08.434-0400	WARN	keep-ecdsa	Please broadcast Bitcoin transaction <transaction-hex>
2021-04-16T11:41:08.434-0400	WARN	keep-ecdsa	Please broadcast Bitcoin transaction <transaction-hex>
```

Broadcast the transaction:
```
$ bitcoind-wallet sendRawTransaction <transaction-hex>
```

And then verify that your fresh wallet address received ~1/30th of a BTC:

```
$ bitcoind-wallet getBalance <address-from-getNewAddress>
0.03329033
```

If these steps are repeated *without* the code in this PR, the `sendRawTransaction` command should return `{"code":-25,"message":"bad-txns-inputs-missingorspent”}` because the `prev_hash` of the transaction is reversed relative to the `txid` of the funding transaction.

tagging @nkuba for review

### CODE REVIEW

File: pkg/chain/tbtc.go [hunk 1]
```diff
@@ -110,9 +110,10 @@ type TBTCSystem interface {
 
 // FundingInfo represents the funding information for a tbtc deposit
 type FundingInfo struct {
-	UtxoValueBytes [8]uint8
-	FundedAt       *big.Int
-	UtxoOutpoint   []uint8
+	UtxoValueBytes  [8]uint8
+	FundedAt        *big.Int
+	TransactionHash string
+	Index           uint32
```
Comment (nkuba) [inline pkg/chain/tbtc.go:116]:
```suggestion
	OutputIndex     uint32
```

File: pkg/chain/local/tbtc.go [hunk 2]
```diff
@@ -185,16 +186,16 @@ func (tlc *TBTCLocalChain) CreateDeposit(
...
 		state:       chain.AwaitingSignerSetup,
 		utxoValue:   big.NewInt(defaultUTXOValue),
 		fundingInfo: &chain.FundingInfo{
-			UtxoValueBytes: utxoValueBytes, // 0x0065cd1d00000000
-			FundedAt:       fundedAt,
-			UtxoOutpoint:   utxoOutpoint, // 0xc27c3bfa8293ac6b303b9f7455ae23b7c24b8814915a6511976027064efc4d5101000000
+			UtxoValueBytes:  utxoValueBytes, // 0x0065cd1d00000000
+			FundedAt:        fundedAt,
+			TransactionHash: previousTransactionHashHex,
+			Index:           previousTransactionIndex,
```
Comment (nkuba) [inline pkg/chain/local/tbtc.go:198]: We need to test the conversion that is done by `FundingInfo` function in `pkg/chain/ethereum/tbtc.go`. Maybe modifying this mock would be too complex, in such case we could add a `pkg/chain/ethereum/tbtc_test.go`  that would cover testing of slicing and byte reversing the `UtxoOutpoint` returned by the ethereum chain.

**Review by nkuba — CHANGES_REQUESTED** [review]

File: pkg/chain/local/tbtc.go (hunk 2 above)
Comment (beaushinkle) [reply pkg/chain/local/tbtc.go:198]: It seems very difficult to properly mock all of the stuff I'd need in order to get a deposit from the chain, and then get the funding info from that deposit. What if I instead pulled the transaction hash / OutputIndex parsing out into a helper function and then unit tested that?

Comment (nkuba) [reply pkg/chain/local/tbtc.go:198]: It sounds very reasonable.

**Comment by beaushinkle** [comment]
good for another look!

File: pkg/chain/ethereum/tbtc.go [hunk 3]
```diff
@@ -466,6 +468,18 @@ func (ta *tbtcApplication) getDepositContract(
 	return depositContract, nil
 }
 
+func parseUtxoOutpoint(utxoOutpoint []uint8) (string, uint32) {
```
Comment (nkuba) [inline pkg/chain/ethereum/tbtc.go:471]: We're duplicating the code and tests in `celo` and `ethereum` packages. We could extract it to some common place. I'm not sure where though. Maybe `pkg/chain/tbtc.go`, or other file under `pkg/chain`?

File: pkg/chain/celo/tbtc.go [hunk 4]
```diff
@@ -451,10 +465,13 @@ func (ta *tbtcApplication) FundingInfo(
 	if err != nil {
 		return nil, err
 	}
+
+	transactionHash, outputIndex := parseUtxoOutpoint(fundingInfo.UtxoOutpoint)
```
Comment (nkuba) [inline pkg/chain/celo/tbtc.go:469]: Let's add an empty line.

```suggestion
	transactionHash, outputIndex := parseUtxoOutpoint(fundingInfo.UtxoOutpoint)
	
```

File: pkg/chain/ethereum/tbtc.go [hunk 5]
```diff
@@ -478,9 +492,12 @@ func (ta *tbtcApplication) FundingInfo(
 	if err != nil {
 		return nil, err
 	}
+
+	transactionHash, outputIndex := parseUtxoOutpoint(fundingInfo.UtxoOutpoint)
```
Comment (nkuba) [inline pkg/chain/ethereum/tbtc.go:496]: Let's add an empty line.

```suggestion
	transactionHash, outputIndex := parseUtxoOutpoint(fundingInfo.UtxoOutpoint)
	
```

**Review by nkuba — CHANGES_REQUESTED** [review]

**Review by nkuba — APPROVED** [review]

**Pull request merged**